spark3/src/main/java/org/apache/iceberg/actions/Spark3CreateAction.java

```diff
@@ -0,0 +1,290 @@
+          throw new RuntimeException(e);
+        } catch (NoSuchNamespaceException noSuchNamespaceException) {
+          Log.error("Cannot recreate source table. Source namespace has been removed, cannot recreate", e);
+          throw new RuntimeException(e);
+        }
+      } catch (TableAlreadyExistsException e) {
+        Log.error("Cannot migrate, Source table was recreated before replacement could be moved. " +
+            "Attempting to remove replacement table.", e);
+        stagingCatalog.dropTable(replacementTable);
+        stagedTable.abortStagedChanges();
+      }
+    }
+
+    return numMigratedFiles;
+  }
+
+  @Override
+  public CreateAction withAdditionalProperties(Map<String, String> properties) {
+    this.additionalProperties.putAll(properties);
+    return this;
+  }
+
+  @Override
+  public CreateAction withAdditionalProperty(String key, String value) {
+    this.additionalProperties.put(key, value);
+    return this;
+  }
+
+  @Override
+  public CreateAction as(String newName) {
+    Spark3Util.CatalogAndIdentifier newDest = Spark3Util.catalogAndIdentifier(spark, newName);
+    return new Spark3CreateAction(spark, sourceCatalog, sourceTableName, newDest.catalog(), newDest.identifier())
+        .withAdditionalProperties(this.additionalProperties);
+  }
+
+  private static void validateSourceTable(CatalogTable sourceTable, Set<String> supportedSourceTableProviders) {
+    String sourceTableProvider = sourceTable.provider().get().toLowerCase(Locale.ROOT);
+
+    if (!supportedSourceTableProviders.contains(sourceTableProvider)) {
+      throw new IllegalArgumentException(
+          String.format("Cannot create an Iceberg table from source provider: %s", sourceTableProvider));
+    }
+    if (sourceTable.storage().locationUri().isEmpty()) {
+      throw new IllegalArgumentException("Cannot create an Iceberg table from a source without an explicit location");
+    }
+  }
+
+  private static Map<String, String> extraIcebergTableProps(String tableLocation, String metadataLocation) {
+    return ImmutableMap.of(
+        TableProperties.WRITE_METADATA_LOCATION, metadataLocation,
+        TableProperties.WRITE_NEW_DATA_LOCATION, tableLocation,
+        "migrated", "true");
+  }
+
+  private static StagingTableCatalog checkDestinationCatalog(CatalogPlugin catalog) {
+    if (!(catalog instanceof SparkSessionCatalog) && !(catalog instanceof SparkCatalog)) {
+      throw new IllegalArgumentException(String.format("Cannot create Iceberg table in non Iceberg Catalog. " +
+              "Catalog %s was of class %s but %s or %s are required", catalog.name(), catalog.getClass(),
+          SparkSessionCatalog.class.getName(), SparkCatalog.class.getName()));
+    }
+    return (StagingTableCatalog) catalog;
+  }
+
+  private CatalogPlugin checkSourceCatalog(CatalogPlugin catalog) {
+    // Currently the Import code relies on being able to look up the table in the session code
```

Comment:
Session catalog?

Do we have a way around this restriction? In our implementation, we load the source table using our Hive or Spark table implementation and check that it is what we expect. Then we use that implementation to load the partitions. Would we similarly require a v2 table implementation to make this catalog agnostic?

(This isn't a blocker, just curious to hear your ideas.)

Reply:
The limitation here is based on how the current SparkUtil.import code is written. It an't handle catalogs or things like that so I thought this would be the easiest approach for now.

I think what we should do is modify that utility function to be less SparkCatalog specific, or at least have ways of calling it that describe a table based on partitions (like in your code) and other properties directly rather than assuming it is a "table name" that can be looked up in the catalog.

There other approach I was thinking about was writing a version of the function which is just all in on Spark3 and uses the CatalogV2 table api. We already have a few discussions going on in issues about how we are going to deal with this for other Actions which all have similar limitations at the moment.

Reply:
I think we can change `SparkTableUtil` we need to.